src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableUpcallHandler.java

```diff
@@ -98,7 +99,7 @@ private void invoke(MemoryAddress buffer) {
                                 ? stackArgsBase.asSlice(storage.index() * abi.arch.typeSize(abi.arch.stackType()))
                                 : bufferBase.asSlice(layout.argOffset(storage));
                             return SharedUtils.read(ptr, type);
-                        });
+                        }, DEFAULT_ALLOCATOR);
```

Comment:
For upcalls, couldn't we use a `NativeScope` and close at the end of the call? If the callback leaks segments outside... well, too bad.

Reply:
Yes, exactly. But, seemed like a separate patch.